Improve performance of recursive

chiro-hiro opened the pull request:
Im try to minimize the cost to `.copy()`, `.clone()` multiple structures around. You might expect ~10% in total performance and ~5% in memory saving.

2 tests are broken, so I'm try to fix it. If this worth, please let me know.

chiro-hiro commented:
@microsoft-github-policy-service agree [company="OROCHI NETWORK PTE. LTD."]

chiro-hiro commented:
@microsoft-github-policy-service agree company="OROCHI NETWORK PTE. LTD."

srinathsetty commented:
@chiro-hiro Is this still a draft? Could you please rebase with the main branch and let me know when this is ready?

chiro-hiro commented:
> @chiro-hiro Is this still a draft? Could you please rebase with the main branch and let me know when this is ready?

Thank you, I will rebase my code and try to complete this in the weekend.

chiro-hiro commented:
Here is your original code

```rust
    // produce a recursive SNARK
    let mut recursive_snark: Option<
      RecursiveSNARK<
        G1,
        G2,
        TrivialTestCircuit<<G1 as Group>::Scalar>,
        CubicCircuit<<G2 as Group>::Scalar>,
      >,
    > = None;

    for _i in 0..num_steps {
      let res = RecursiveSNARK::prove_step(
        &pp,
        recursive_snark,
        circuit_primary.clone(),
        circuit_secondary.clone(),
        vec![<G1 as Group>::Scalar::ONE],
        vec![<G2 as Group>::Scalar::ZERO],
      );
      assert!(res.is_ok());
      recursive_snark = Some(res.unwrap());
    }

    assert!(recursive_snark.is_some());
    let recursive_snark = recursive_snark.unwrap();

    // verify the recursive SNARK
    let res = recursive_snark.verify(
      &pp,
      num_steps,
      vec![<G1 as Group>::Scalar::ONE],
      vec![<G2 as Group>::Scalar::ZERO],
    );
    assert!(res.is_ok());
```

Instead of create new instance of `RecursiveSNARK` every time, I changed the code to manage a single mutable instance of `RecursiveSNARK`.

```rust
    // produce a recursive SNARK
    let mut recursive_snark = RecursiveSNARK::<
      G1,
      G2,
      TrivialTestCircuit<<G1 as Group>::Scalar>,
      CubicCircuit<<G2 as Group>::Scalar>,
    >::new(
      &pp,
      &circuit_primary,
      &circuit_secondary,
      vec![<G1 as Group>::Scalar::one()],
      vec![<G2 as Group>::Scalar::zero()],
    );

    for _i in 0..num_steps {
      let res = recursive_snark.prove_step(
        &pp,
        &circuit_primary,
        &circuit_secondary,
        vec![<G1 as Group>::Scalar::one()],
        vec![<G2 as Group>::Scalar::zero()],
      );
      assert!(res.is_ok());
    }

    // verify the recursive SNARK
    let res = recursive_snark.verify(
      &pp,
      num_steps,
      vec![<G1 as Group>::Scalar::ONE],
      vec![<G2 as Group>::Scalar::ZERO],
    );
    assert!(res.is_ok());
```

I observed, the circuit and z0 were passed without mutable. I changed the code to borrow these values instead of `.clone()` or copy.

srinathsetty commented on src/lib.rs:
```diff
@@ -193,197 +195,212 @@ where
-          None,
-        );
+    // Frist step was already done in the constructor
+    if self.i == 0 {
```
Does it make sense to set i to 1 in the constructor?

chiro-hiro replied:
I tried, the problem is wen I do this we might need to change the entire code to staring prove from step 1.

```rust
for i in 1..step_numbers {
  recursive_snark.prove_step(...);
}
```

srinathsetty replied:
Good point! Let me look at the code again and see if there is a way to fix this without having to change the step numbers.

srinathsetty replied:
@chiro-hiro @huitseeker After thinking more, I think changing prove_step to only prove from 1..step_numbers might be the cleanest option. Could we do this?

Also, we just merged a PR from @huitseeker to address clippy issues. Can we please rebase this branch?

huitseeker replied:
I've put up a commit to sketch what that change would look like (don't use as-is, it contains a couple of commented lines that should be cleaned up):
https://github.com/huitseeker/Nova/commit/6342a263b80a347a4e3a252b88bc8dcb4037eca7

I'm not convinced that changing proof iterations to run [1..n] is more ergonomic than the idempotent / backwards-compatible approach of the present PR, which makes miscounting steps hard. But YMMV - and I hope the commit is directionally useful if @srinathsetty thinks this change should happen.

chiro-hiro replied:
I'm prefer to start from `0` instead of `1`. If we want to solve this completely we might need to change a lot of code , and change the `RecursiveSNARK`'s context to optional `Option<RecursiveSNARKContext>` to allow us init a new instance without prove anything.

srinathsetty replied:
Thanks, @huitseeker and @chiro-hiro! We can keep the idempotent approach in the present PR.

huitseeker replied:
> If we want to solve this completely we might need to change a lot of code , and change the RecursiveSNARK's context to optional Option<RecursiveSNARKContext> to allow us init a new instance without prove anything.

I think it's an interesting idea, that may be worth opening an issue @chiro-hiro : it's also something we can perhaps engineer with a `OnceCell`, especially since they're now stabilized in Rust 1.70.

srinathsetty commented:
Fantastic! Thanks @chiro-hiro! I had a quick comment about avoiding a small piece of code in prove_step. Please check.

chiro-hiro commented:
My pleasure,

Indeed, that part of code looks ugly but I can't find the better way to keep the backwards compatibility.

chiro-hiro commented:
Hi @srinathsetty, there are pretty much changes was made in `main` branch. It's a bit overwhelming to keep my branch compatible. Could we speed up the process for this PR?.

srinathsetty commented:
> Hi @srinathsetty, there are pretty much changes were made in `main` branch. It's a bit overwhelming to keep my branch compatible. Could we speed up the process for this PR?.

Hi @chiro-hiro! I'll see if I can find a fix for the issue noted above. We'll prioritize this PR.

chiro-hiro commented:
Thank you, I will focus on fixing the conflict with the main branch.

chiro-hiro commented:
@huitseeker Thank you, everything looks great. I do check and rerun the test/benchmark as well. All set, just wait for @srinathsetty to review this PR.

huitseeker commented:
@chiro-hiro warnings should disappear as soon as you rebase on the latest master.

chiro-hiro commented:
Hi @srinathsetty, sorry for delay. I have sync my fork with the upstream's main branch.

chiro-hiro commented:
Hi @srinathsetty, I rebased my code. Please help me check!.

huitseeker commented:
@chiro-hiro 

> Should I fix this?

That's really a question for @srinathsetty, but yeah, I suspect so. Apologies for overlooking this in prior PRs.

huitseeker commented:
@chiro-hiro 
> Should I fix this?

That's really a question for @srinathsetty but yes, I suspect so. Apologies for overlooking this in earlier PRs.

huitseeker commented on Cargo.toml:
```diff
@@ -37,7 +37,7 @@ thiserror = "1.0"
 
 [dev-dependencies]
-criterion = "0.3.1"
+criterion = { version = "0.5.1", features = ["html_reports"] }
```
Version 0.5.1 is not yet compatible pprof, which I introduce in #183. It would be great to only upgrade to 0.4 for now.

chiro-hiro replied:
I see, thank you.

huitseeker commented:
@chiro-hiro I've opened https://github.com/orochi-network/Nova/pull/2 which is a minimal fix of [1](https://github.com/microsoft/Nova/pull/163#issuecomment-1594079298) and [2](https://github.com/microsoft/Nova/pull/163#discussion_r1232439399).
